includes/class-newspack-ads-settings.php

```diff
@@ -0,0 +1,324 @@
+	 */
+	public static function api_permissions_check() {
+		if ( ! current_user_can( self::API_CAPABILITY ) ) {
+			return new \WP_Error(
+				'newspack_ads_rest_forbidden',
+				esc_html__( 'You cannot use this resource.', 'newspack-ads' ),
+				[
+					'status' => 403,
+				]
+			);
+		}
+		return true;
+	}
+
+	/**
+	 * Sanitize settings coming from the API.
+	 *
+	 * @param array           $settings Settings to sanitize.
+	 * @param WP_REST_Request $request  Full details about the request.
+	 *
+	 * @return array Sanitized settings.
+	 */
+	public static function api_sanitize_settings( $settings, $request ) {
+		$section            = (string) $request['section'];
+		$sanitized_settings = [];
+		foreach ( $settings as $key => $value ) {
+			$config = self::get_setting_config( $section, $key );
+			settype( $value, $config['type'] );
+			$sanitized_settings[ $key ] = $value;
+		}
+		return $sanitized_settings;
+	}
+
+	/**
+	 * Update setting section.
+	 *
+	 * @return WP_REST_Response containing the settings list.
+	 */
+	public static function api_get_settings_list() {
+		return \rest_ensure_response( self::get_settings_list() );
+	}
+
+	/**
+	 * Update setting section.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response containing the settings list.
+	 */
+	public static function api_get_settings( $request ) {
+		return \rest_ensure_response( self::get_settings( $request['section'] ) );
+	}
+
+	/**
+	 * Update setting section.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response containing the settings list.
+	 */
+	public static function api_update_section( $request ) {
+		return \rest_ensure_response( self::update_section( $request['section'], $request['settings'] ) );
+	}
+
+	/**
+	 * Get the setting option name to be used on the options table.
+	 *
+	 * @param object $setting The setting object to retrieve the key from.
+	 *
+	 * @return string Option name. 
+	 */
+	private static function get_setting_option_name( $setting ) {
+		return self::OPTION_NAME_PREFIX . $setting['section'] . '_' . $setting['key'];
+	}
+
+	/**
+	 * Retrieves list of configured settings.
+	 *
+	 * A setting is an array with the following keys:
+	 * - description: The description of the setting.
+	 * - help: The help text for the setting.
+	 * - section: The section the setting is in.
+	 * - key: The key of the setting. Should be used along with the section name.
+	 * - type: The type of the setting. Used to typecast the value.
+	 * - default: The default value of the setting.
+	 * - options: Options to be used for a select field. Values outside of this array will not update.
+	 *   - name: The name of the option.
+	 *   - value: The value of the option.
+	 * - public: Whether the setting value is allowed to be displayed publicly on the frontend.
+	 *
+	 * Settings without `key` or with the `key` of value `active` should be
```

Comment:
I really like the idea of creating a more standardized way of structuring settings config for use by the Newspack wizard UI! I wonder if we could eventually port this to the main plugin and use it across all of our repos. Before that, maybe the API warrants some discussion?

For example, WP Core already has the concept of [settings sections](https://developer.wordpress.org/reference/functions/add_settings_section/) which are basically just configured with a slug and a display title—why not mimic this API instead of using a flat array structure and "special" or missing key values to denote the section? Might also make it a bit easier to group the settings by section up-front (as returned by the API endpoint), instead of handling it in JS using [`groupBy`](https://github.com/Automattic/newspack-plugin/pull/1208/files#diff-bde388cd74998cc3a84c1b3330379f7350c1dc99d7b73ee185833495fbdf39a5R66).

Reply:
I thought about this too! After testing different scenarios and different plugins, it can be part of a plugin module registration logic.

I'm also taking suggestions on its model. My thoughts on the section system are towards simplification, we are currently not *registering* these, like the WP Core implementation. [As you can see how the UI currently  implements it](https://github.com/Automattic/newspack-plugin/blob/eb39b664dab93af99dcb7114cc6854d0f664361f/assets/wizards/advertising/views/settings/settings-section/index.js), this solution handles well when a section is a "module" that can be activated. If we decide that the `groupBy` after fetch is not ideal, the grouping can live on the backend. Perhaps a different API endpoint or an endpoint argument. WDYT?

Reply:
d9c8e6518decf1f95d0dd17c333fc7726d1499b6 along with https://github.com/Automattic/newspack-plugin/pull/1208/commits/4ab2f2f3036f69d9461febd577b72617d59ea741 implements associative array support on the `get_settings_list()` method and use as default for the API endpoint

Reply:
I get you. One idea for shadowing the core settings section API is that it might allow us to "detach" these settings from the main plugin/wizard UI, so that if we don't have the wizard, we could always display a more Core-friendly settings page using the same config arrays. But since that might not be the direction we go, I'm okay keeping the current model for the simplicity.

Reply:
Thank you for your considerations! I'm happy to revisit this once we start implementing it in other plugins. Even though it is a structural part of the concept, it doesn't affect how data is stored and shouldn't require too much rework and no backward compatibility to adapt to a different approach.

How I'm seeing it now is that we can get the configuration from `get_settings_list()` and build the core friendly `add_settings_section()` and `add_settings_field()` in the case of a standalone setup while we work our way into modularizing the wizard js so it can live in the plugin's repository.

Although it could work the other way around, I like the idea of setting the parameters ourselves so it can better communicate with how the wizard and its components are put together.

Reply:
Sounds good to me—I do appreciate the value of structuring it for the primary use case (the wizard UI) rather than the fallback (core WP settings). Thanks for talking it through!